### ISSUE TITLE
Incorrect Reporting of Invalid Private Attribute Access when Self-Narrowing

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#9596]
**Describe the bug**
When you have a type-guard on `self` as being a subclass, in the positive case, Pyright complains about private attribute access despite the attribute still belonging to the class. Let's take this example:

```python

class _MyBase:
    _private_attr = 5

    def foo(self):
        if isinstance(self, Base1):
            print(self._private_attr) # "_private_attr" is protected and used outside of the class in which it is declared
        else:
            print(self._private_attr)

class Base1(_MyBase):
    _private_attr = 10
```

[Pyright Playground
](https://pyright-play.net/?pyrightVersion=1.1.390&pythonVersion=3.13&strict=true&code=MYGwhgzhAED6CyBPAQpApgLgLAChr7gAcAnASwDcwAXNWaq46AXmgFZdcDoATNAM2h8A9kIAUENCD4BKbHi4FSA0hFIA7CFTBrgacZL4AaaKgkBGWZwXWS6qvqkA6WLco06VBtKvXJEudY2ZGr2Ek4uZG609MTeOLigkDCmaGaiCCjolvL4ERTU0Z6MLGYADLhAA)

**VS Code extension or command-line**
Pyright 1.1.390

**Comment by erictraut** [comment]
I don't think I'd consider this a bug.

What you're doing in this code sample violates a basic rule of object-oriented programming: a base class should not have knowledge of its child classes. Child classes are allowed to access private members of their parent classes, but the opposite should not be true. The `reportPrivateUsage` enforces this.

I'll also note that `reportPrivateUsage` is enabled by default only in strict mode. Strict-mode checks tend to be more opinionated in how they work.

My recommendation is to look for ways to redesign this code to avoid this condition. If you want to keep it as is, you can suppress the `reportPrivateUsage` in this instance by using a `# type: ignore`.

**Comment by max-muoto** [comment]
> I don't think I'd consider this a bug.
> 
> What you're doing in this code sample violates a basic rule of object-oriented programming: a base class should not have knowledge of its child classes. Child classes are allowed to access private members of their parent classes, but the opposite should not be true. The `reportPrivateUsage` enforces this.
> 
> I'll also note that `reportPrivateUsage` is enabled by default only in strict mode. Strict-mode checks tend to be more opinionated in how they work.
> 
> My recommendation is to look for ways to redesign this code to avoid this condition. If you want to keep it as is, you can suppress the `reportPrivateUsage` in this instance by using a `# type: ignore`.

Makes sense, thanks Eric!
